fs/background-fetcher/background_fetcher.go

```diff
@@ -0,0 +1,162 @@
+/*
+   Copyright The Soci Snapshotter Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package backgroundfetcher
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	sm "github.com/awslabs/soci-snapshotter/fs/span-manager"
+	"github.com/awslabs/soci-snapshotter/soci"
+	"github.com/containerd/containerd/log"
+	"github.com/opencontainers/go-digest"
+	"golang.org/x/time/rate"
+)
+
+type layerResolver struct {
+	*sm.SpanManager
+	layerDigest     digest.Digest
+	nextSpanFetchID soci.SpanID
+}
+
+type Option func(*BackgroundFetcher) error
+
+const (
+	// Amount of time the background fetcher will wait once a new layer comes in
+	// before (re)starting fetches.
+	defaultSilencePeriod = 5 * time.Second
+
+	// The minimum size of data to be fetched in a single fetch.
+	defaultMinFetchSize = 1 << 22 // 4 MiB
+
+	// Specifies how often the fetch will occur.
+	// The background fetcher will fetch a single span every `defaultFetchPeriod`.
+	defaultFetchPeriod = 500 * time.Millisecond
+)
+
+func WithMinFetchSize(size int64) Option {
+	return func(bf *BackgroundFetcher) error {
+		bf.minFetchSize = size
+		return nil
+	}
+}
+
+func WithSilencePeriod(period time.Duration) Option {
+	return func(bf *BackgroundFetcher) error {
+		bf.silencePeriod = period
+		return nil
+	}
+}
+
+func WithFetchPeriod(period time.Duration) Option {
+	return func(bf *BackgroundFetcher) error {
+		bf.fetchPeriod = period
+		return nil
+	}
+}
+
+// A backgroundFetcher is responsible for fetching spans from layers
+// in the background.
+type BackgroundFetcher struct {
```

Comment:
Is this the right abstraction? The BackgroundFetcher handles both work queueing, but also the state of the work to be done. Could we separate it into a WorkQueue/JobQueue and a job that actually knows how to do a single unit of background fetching? Is there an existing implementation of a queue that we could use?

Reply:
Hmm, does it make sense to do something like 


```
type Resolver interface {
    Resolve(ctx context.Context) error 
}

func (lr *layerResolve) Resolve() error {
    // fetch from data
}

...

type BackgroundFetcher struct {
     ...
     workQueue chan Resolver
}


func (bf *BackgroundFetcher) Run() error {
   ...
   select {
        ...
        case lr := <- workQueue:
                  err := lr.ResolveNext()
                  if err == nil {
                      workQueue <- lr
                   }
       }
 }
```

Reply:
I think that makes sense. It could make it easier for us to, for example, add an optional LOD for a layer to the bgfetcher later.

Reply:
Done.